merkle_tree/src/main.rs

```diff
@@ -0,0 +1,432 @@
+    fn new(data_vector: Vec<&str>) -> Self {
+        let string_data_vec: Vec<String> = data_vector.iter().map(|s| s.to_string()).collect();
+        let pow2_data_vector = extend_to_power2_size(&string_data_vec);
+        let mut level_n = vec![];
+        for data in &pow2_data_vector {
+            level_n.push(digest(data.clone()));
+        }
+        let mut current_tree = vec![];
+        for level in 0..(pow2_data_vector.len() as f64).sqrt().ceil() as usize {
+            //Amount of levels
+            let mut level_n_minus1 = vec![];
+            for i in (0..level_n.len()).step_by(2) {
+                //Build new level from the previous one
+                level_n_minus1.push(digest(level_n[i].clone() + &level_n[i + 1]));
+            }
+            let next_current_level = level_n_minus1.clone();
+            if level == 0 {
+                level_n_minus1.append(&mut level_n); //Concatenating the leaves.
+            } else {
+                level_n_minus1.append(&mut current_tree); //Concatenating the previous levels.
+            }
+            current_tree = level_n_minus1.clone();
+            level_n = next_current_level;
+        }
+
+        Self {
+            tree: current_tree.clone(),
+            leaves: data_vector.len(),
+            og_leaves: string_data_vec.clone(),
+        }
+    }
+
+    /// Given a proof and an index of a data hash the function must determine wether the proof is correct.
+    /// For understanding this function we can consider the following tree:
+    ///               root
+    ///          h12        h34
+    ///        h1   h2    h3   h4
+    ///        |    |     |     |
+    ///     data1 data2 data3 data4
+    /// The array representation would be: [root,h12,h34,h1,h2,h3,h4]
+    /// With the indexes                     0   1   2   3  4  5   6
+    fn verify(&self, proofs: Vec<String>, mut leaf_index: usize) -> bool {
+        let mut hash_for_verification = self.tree[leaf_index].clone();
+        for proof in proofs {
+            if leaf_index % 2 == 1 {
+                // if index is odd we are on a left branch, so the verification must be computed concatenating the proof second
+                hash_for_verification = digest(hash_for_verification.to_owned() + &proof);
+            } else {
+                // if index is even we are on a right branch, so the verification must be computed concatenating the proof first
+                hash_for_verification = digest(proof.clone() + &hash_for_verification);
+                leaf_index -= 1; //if it's a right child this is necesary to calculate its parent
+            }
+            leaf_index = leaf_index / 2;
+        }
+        hash_for_verification == self.tree[0]
+    }
+
+    /// Given an index of a data hash the function must return the proof that the tree contains that data hash.
+    fn generate_proof(&self, mut node_index: usize) -> Option<Vec<String>> {
```

Comment:
This function is a good place to use the `.get(index)` function instead of brackets indexing. get will return an option that can be instantly returned with the "?" operator (try operator). This will make the function return None if get returns None, or continue with the execution. For example, you can do: `self.tree.get(node_index + 1)?`